locator/src/control_plane.rs

```diff
@@ -39,120 +37,117 @@ pub enum ControlPlaneError {
+        const RETRIABLE_STATUS_CODES: &[StatusCode] = &[
+            StatusCode::TOO_MANY_REQUESTS,     // 429
+            StatusCode::INTERNAL_SERVER_ERROR, // 500
+            StatusCode::BAD_GATEWAY,           // 502
+            StatusCode::SERVICE_UNAVAILABLE,   // 503
+            StatusCode::GATEWAY_TIMEOUT,       // 504
+        ];
+
+        let mut cells: HashMap<String, Arc<Cell>> = HashMap::new();
+        let mut org_to_cell = HashMap::new();
+        let mut next_cursor: Option<String> = cursor.map(String::from);
+        let mut page_fetches = 0;
+
+        // 3 retries per page fetch
+        let mut retries = 0;
+
+        loop {
+            let mut url = Url::parse(&self.full_url)
+                .map_err(|e| ControlPlaneError::InvalidUrl(e.to_string()))?;
+
+            if let Some(ref c) = next_cursor {
+                url.query_pairs_mut().append_pair("cursor", c);
+            }
+
+            let response = self.client.get(url).send().await?;
+
+            if !response.status().is_success() {
+                if RETRIABLE_STATUS_CODES.contains(&response.status()) && retries < 3 {
+                    // Backoff between retries
+                    let retry_millis = BASE_DELAY * 2_u64.pow(retries);
+                    sleep(Duration::from_millis(retry_millis)).await;
+                    retries += 1;
+                    continue;
+                } else {
+                    return Err(ControlPlaneError::ControlPlaneRetriesExceeded);
+                }
             }
-        }
 
-        // Response successful, reset retries counter
-        retries = 0;
+            // Response successful, reset retries counter
+            retries = 0;
 
-        let json_response = response.json::<ControlPlaneData>().await?;
+            let json_response = response.json::<ControlPlaneData>().await?;
 
-        for (c, l) in json_response.metadata.cell_to_locality {
-            cells.insert(c.clone(), Arc::new(Cell::new(c, l)));
-        }
+            for (c, l) in json_response.metadata.cell_to_locality {
+                cells.insert(c.clone(), Arc::new(Cell::new(c, l)));
+            }
 
-        for row in json_response.data {
-            org_to_cell.insert(row.id, row.cell.clone());
-            org_to_cell.insert(row.slug, row.cell);
-        }
+            for row in json_response.data {
+                org_to_cell.insert(row.id, row.cell.clone());
+                org_to_cell.insert(row.slug, row.cell);
+            }
 
-        page_fetches += 1;
-        next_cursor = Some(json_response.metadata.cursor);
+            page_fetches += 1;
+            next_cursor = Some(json_response.metadata.cursor);
 
-        if !json_response.metadata.has_more {
-            break;
+            if !json_response.metadata.has_more {
+                break;
+            }
         }
-    }
 
-    println!("Fetched {} pages from control plane", page_fetches);
+        println!("Fetched {} pages from control plane", page_fetches);
```

Comment:
Would this be better as a `tracing::debug()` log event?

Reply:
yeah sorry, let me look into these next... all the println!s right now are placeholders so we'll remember to add tracing and metrics in these spots